<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Selecting an optimal number of degrees of freedom • santaR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Selecting an optimal number of degrees of freedom">
<meta property="og:description" content="santaR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">santaR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/getting-started.html">Intro</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting-started.html">Getting started</a>
    </li>
    <li>
      <a href="../articles/prepare-input-data.html">How to prepare input for santaR</a>
    </li>
    <li>
      <a href="../articles/santaR-GUI.html">Graphical User Interface</a>
    </li>
    <li>
      <a href="../articles/theoretical-background.html">santaR theoretical background</a>
    </li>
    <li>
      <a href="../articles/selecting-optimal-df.html">Selecting an optimal number of degrees of freedom</a>
    </li>
    <li>
      <a href="../articles/automated-command-line.html">Automated command line analysis</a>
    </li>
    <li>
      <a href="../articles/plotting-options.html">Plotting options</a>
    </li>
    <li>
      <a href="../articles/advanced-command-line-functions.html">Advanced command line functions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/adwolfer/santaR" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Selecting an optimal number of degrees of freedom</h1>
                        <h4 data-toc-skip class="author">Arnaud Wolfer</h4>
            
            <h4 data-toc-skip class="date">2019-10-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/adwolfer/santaR/blob/HEAD/vignettes/selecting-optimal-df.Rmd" class="external-link"><code>vignettes/selecting-optimal-df.Rmd</code></a></small>
      <div class="hidden name"><code>selecting-optimal-df.Rmd</code></div>

    </div>

    
    
<p><code>santaR</code> is a Functional Data Analysis (<em>FDA</em>) approach where each individual’s observations are condensed to a continuous smooth function of time, which is then employed as a new analytical unit in subsequent data analysis.</p>
<p>Fitting a smooth function to a time trajectory is equivalent to a denoising or signal extraction problem; a balance between the fitting of the raw data and the smoothing of the measurement error (or biological variability) must be found (see <a href="theoretical-background.html">santaR theoretical background</a>).</p>
<p><code>santaR</code> parametrise the smoothness of individual smooth-splines fit by fixing the number of effective degrees of freedom allowed (<em>df</em>) . <em>df</em> can be described as the single meta-parameter that a user must select. Although the ‘ideal’ solution, automated approaches for the selection of an optimal level of smoothing are still an area of active research; and while no definitive answer on the most suitable methodology can be provided, the following vignette will present the current state of these approaches and propose a strategy to help a user make an informed decision.</p>
<p>As the true underlying function of time is inaccessible and the optimal <em>df</em> unknown, strategies for the estimation of smoothness and model selection must be devised. The fundamental question that must guide the tuning of the smoothing parameter is to know which observed curve features are <em>“real”</em> and which are spurious artefacts of the fitting procedure (over-fitted). Most <em>FDA</em> algorithms rely on an automated selection of the smoothness parameter; yet as smoothness is central to the denoising and signal extraction procedure, the quality of the results will be highly dependent on the tuning<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. It can be noted that this challenge is shared with multiple other data analysis methods imposing a form of smoothness. The problem of selecting the window width or bandwidth in a kernel density estimator (KDE) is notably similar, resulting in shared literature and strategies<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> <a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> <a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>; even if most reported algorithm require around 20 time-points while most short trajectories hardly reach half this number.</p>
<p>The following vignette will focus on:</p>
<ul>
<li>Intuitive parametrisation of smoothing</li>
<li>Automated model assessment and selection approaches</li>
<li>Latent time-trajectories for <em>df</em> selection</li>
<li>Smoothness and experimental design</li>
</ul>
<div class="section level2">
<h2 id="intuitive-parametrisation-of-smoothing">Intuitive parametrisation of smoothing<a class="anchor" aria-label="anchor" href="#intuitive-parametrisation-of-smoothing"></a>
</h2>
<p>Based on simulated data and diverse datasets, some intuitive rules for the selection of <em>df</em> can be established:</p>
<blockquote>
<ul>
<li><p><em>df</em> controls the <em>“complexity”</em> of the model employed. A substantial difference can be found when going from <em>2</em> to <em>10</em>, but very little change will take place when going from <em>10</em> to <em>50</em> (the model only gets more complex, but the general shape won’t change).</p></li>
<li><p>More time points do not automatically require a higher <em>df</em>. More inflexions (more complex shape) could require a higher <em>df</em> if the number of points is sufficient (and the sampling frequency high).</p></li>
<li><p>A lower <em>df</em> value is often more suited and generalisable (less over-fitted).</p></li>
<li><p>If the <em>df</em> is for example <em>10</em>, all individuals trajectories with less of 10 time-points cannot be fitted and will be rejected.</p></li>
<li><p>On simulated data, the results (<em>p</em>-values) are resilient to most values of <em>df</em>, however the plots can look dramatically different.</p></li>
<li>
<p>Trying multiple values of <em>df</em> on a subset of variables (using the GUI) and then selecting the fit that approximate the time evolution the best without over-fitting:</p>
<ul>
<li>
<em>df=5</em> is a good starting point in most cases (even more so if there is less than 10 time-points)</li>
<li>If the number of time-points is large and the curves seem very under-fitted, <em>df</em> can be increased to <em>6</em>, <em>7</em> or more. Values higher than <em>10</em> should rarely be required and will provide with a diminishing return. <em>df=number of time-points</em> will result in a curve passing through all points (over-fitted).</li>
<li>If the number of points is lower or the trajectories seem over-fitted, <em>df</em> can be decreased to <em>4</em> or <em>3</em>. (<em>3</em> will be similar to a second degree polynomial, while <em>2</em> will be a linear model)</li>
<li>If the plots <em>“look right”</em> and don’t seem to <em>“invent”</em> information between measured data-points, the <em>df</em> is close to optimal.</li>
</ul>
</li>
</ul>
</blockquote>
<p>As will be demonstrated in the following sections (<em>Automated model assessment and selection approaches</em>), it does not seem to be possible to automatically select the degree of freedom. A choice based on visualisation of the splines while being careful with over-fitting, keeping in mind the <em>“expected”</em> evolution of the underlying process seems the most reasonable approach.</p>
<p>In practice, the results are resilient to the <em>df</em> value selected across all variables, which is a function of the <em>study design</em>, such as the <em>number of time-points</em>, <em>sampling rate</em> and, most importantly, the <em>complexity of the underlying function of time</em>. While an automated approach cannot infer the study design from a limited set of observations, an informed user will intuitively achieve a more consistent fit of the data (see <em>Smoothness and Experimental Design</em>).</p>
</div>
<div class="section level2">
<h2 id="automated-model-assessment-and-selection-approaches">Automated model assessment and selection approaches<a class="anchor" aria-label="anchor" href="#automated-model-assessment-and-selection-approaches"></a>
</h2>
<p>In classical parametric statistics, a reduction of the RSS ensures the “closest” fit of the observations. However such a least-square fitting approach cannot be employed as the most over-fitting model would always be selected, resulting in an interpolation of the observations.</p>
<blockquote>
<p>As the present measurements are noisy and the models considered flexible, no “truth” or certainty can be found in the data. This bias set smoothing aside from classical parametric approaches and result in the complexity of automated selection of smoothness.</p>
</blockquote>
<p>The most common methodology to tune the smoothing parameter consists in evaluating a metric (a measure the quality of the model) at different smoothing parameter values, and select the model that minimises/maximise the target metric. These metrics, like the fitting of a smoothingspline itself (see <a href="theoretical-background.html">santaR theoretical background</a> <em>Equation 1</em>), must impose a trade-off between the goodness-of-fit and a penalty on the model complexity. Each metric will therefore potentially present a (different) bias depending on how the balance is established. This bias could ultimately define the metric’s behaviour when a high or low number of samples are available. To evaluate a smoothing parameter for under- or over-fitting, the main model selection procedures employ the cross-validation score (<strong>CV</strong>), general cross-validation score (<strong>GCV</strong>), Akaike information criterion (<strong>AIC</strong>), Bayesian information criterion (<strong>BIC</strong>), or the corrected AIC (<strong>AICc</strong>).</p>
<p>Ordinary <strong>CV</strong> consists in dividing the <em>N</em> data-points in <em>K</em> groups or folds. Each fold will successively be used as a test set, while a model is trained on the remaining data. After a model is fitted (for a given smoothing parameter) with the K<sup>th</sup> fold removed, the prediction error can be evaluated on the (unseen) K<sup>th</sup> fold (test set). When this procedure has been repeated for all folds, each data point as been employed once as a test, and the prediction error across all folds can be averaged. This average prediction error provides an estimate of the curve’s test error for a given smoothness parameter; the parameter that minimises the test error is subsequently selected<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>.</p>
<p><strong>GCV</strong><a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a> is a faster approximation of <strong>CV</strong> relying on the fact that, in some cases, the trace of the roughness matrix can be calculated more easily than the individual elements of its diagonal. <strong>GCV</strong> has been reported to possibly reduce the tendency of <strong>CV</strong> to under-smooth in some condition<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a>. <strong>GCV</strong> and <strong>CV</strong> scores (i.e. test errors) have been described as changing slowly as <span class="math inline">\(\lambda\)</span> approaches the minimising value (corresponding to a wide range of <span class="math inline">\(\lambda\)</span> resulting in close test error values), making optimal smoothness hard to define <a href="#fn8" class="footnote-ref" id="fnref8"><sup>8</sup></a>.</p>
<p>The remaining three metrics commonly employed in model selection (i.e. <strong>AIC</strong>, <strong>BIC</strong>, <strong>AICc</strong>) are less computationally intensive as they rely on the log-likelihood of the model<a href="#fn9" class="footnote-ref" id="fnref9"><sup>9</sup></a> <a href="#fn10" class="footnote-ref" id="fnref10"><sup>10</sup></a>. These metrics penalise models based on the number of parameters, to balance the goodness of fit versus model complexity. The Akaike information criterion<a href="#fn11" class="footnote-ref" id="fnref11"><sup>11</sup></a> <a href="#fn12" class="footnote-ref" id="fnref12"><sup>12</sup></a>, defined as AIC = <span class="math inline">\(2k - 2\mathcal{L}\)</span> , where <span class="math inline">\(\mathcal{L}\)</span> is the log-likelihood and <span class="math inline">\(k\)</span> the number of parameters, rewards goodness of fit (estimated by <span class="math inline">\(\mathcal{L}\)</span>) while increasingly penalising each added parameter. Across candidate models (e.g. varying smoothing parameter employed), the model which minimises <strong>AIC</strong> is preferred.</p>
<p>The Bayesian information criterion<a href="#fn13" class="footnote-ref" id="fnref13"><sup>13</sup></a>, defined as BIC = <span class="math inline">\(log(n)k - 2\mathcal{L}\)</span>, where <span class="math inline">\(n\)</span> is the sample size<a href="#fn14" class="footnote-ref" id="fnref14"><sup>14</sup></a> <a href="#fn15" class="footnote-ref" id="fnref15"><sup>15</sup></a> <a href="#fn16" class="footnote-ref" id="fnref16"><sup>16</sup></a>, is an information criterion established in a Bayesian framework, taking the sample size into account when penalising a model fit. Like <strong>AIC</strong>, the model minimising <strong>BIC</strong> is preferred. Different behaviour of <strong>AIC</strong> and <strong>BIC</strong> can be observed as the number of time-points is altered. As <span class="math inline">\(N \longrightarrow 0\)</span>, <strong>BIC</strong> will select models that are too simple due to the heavy penalty on complexity. Conversely, as <span class="math inline">\(N \longrightarrow \infty\)</span>, <strong>AIC</strong> present a tendency to select models that are too complex while <strong>BIC</strong> present an increasing probability of selecting the true model as the number of samples increases<a href="#fn17" class="footnote-ref" id="fnref17"><sup>17</sup></a>.</p>
<p>In order to address possible limitations of <strong>AIC</strong> for small samples size, Hurvich and Tsai<a href="#fn18" class="footnote-ref" id="fnref18"><sup>18</sup></a>] proposed a corrected <strong>AIC</strong> (<strong>AICc</strong>) defined as AICc = <span class="math inline">\(-2\mathcal{L} + 2k + (2k(k+1)/(n-k-1)\)</span>, when the models are univariate, linear and have normally-distributed residuals. As the sample size increases, <strong>AICc</strong> tends towards <strong>AIC</strong>.</p>
<p>No clear choice between <strong>CV</strong>, <strong>GCV</strong>, <strong>AIC</strong>, <strong>AICc</strong> or <strong>BIC</strong> does exist.</p>
<blockquote>
<p>Bar-Joseph <em>et al</em>.<a href="#fn19" class="footnote-ref" id="fnref19"><sup>19</sup></a> <a href="#fn20" class="footnote-ref" id="fnref20"><sup>20</sup></a> parametrise the smoothing as part of the optimisation problem. EDGE<a href="#fn21" class="footnote-ref" id="fnref21"><sup>21</sup></a> employs <strong>GCV</strong>, SME<a href="#fn22" class="footnote-ref" id="fnref22"><sup>22</sup></a> optimises <strong>AICc</strong> but offer the possibility of <strong>AIC</strong> and <strong>BIC</strong>, while FDA<a href="#fn23" class="footnote-ref" id="fnref23"><sup>23</sup></a> relies on <strong>CV</strong> and Déjean <em>et al</em>.<a href="#fn24" class="footnote-ref" id="fnref24"><sup>24</sup></a> on <strong>GCV</strong>.</p>
</blockquote>
<p>A comparative study between these metrics showed that none uniformly outperformed the others<a href="#fn25" class="footnote-ref" id="fnref25"><sup>25</sup></a>. This simulation, as most successful application of automated selection, employs a high number of data-points per trajectory. If more than 20 observations are available, <strong>CV</strong> and <strong>GCV</strong> can provide a reliable fit of the data; however most metabolic and gene expression trajectories seldom reach half this value. In practice, most FDA-derived approaches tailored for these datasets have reported critical limitations in the automated fitting procedure, with Ramsay <em>et al</em>. (employing <strong>GCV</strong>) describing it as <em>“by no mean foolproof”</em><a href="#fn26" class="footnote-ref" id="fnref26"><sup>26</sup></a> <a href="#fn27" class="footnote-ref" id="fnref27"><sup>27</sup></a>, highlighting the fact that automated smoothing selection is still an area of research. More critically, while it could be expected that different metabolites or genes should be modelled with different complexity (<em>as different underlying processes could govern the value of different variables</em>), in practice the automated selection of a different degree of smoothness for each variable can lead to an <em>“inflation of significance due to overfitting”</em><a href="#fn28" class="footnote-ref" id="fnref28"><sup>28</sup></a> in EDGE, confirmed by the work of Déjean <em>et al</em>.<a href="#fn29" class="footnote-ref" id="fnref29"><sup>29</sup></a> where <em>“tuning the smoothing parameter is a core problem that could not be achieved by the usual cross-validation method because of the poor quality of clustering results”</em>. To counter this challenge, statistical approaches have elected to parametrise a single smoothing value shared by all variables. As a result the modelling approach assumes that all functional curves arise from the same underlying function of time.</p>
</div>
<div class="section level2">
<h2 id="latent-time-trajectories-for-df-selection">Latent time-trajectories for <em>df</em> selection<a class="anchor" aria-label="anchor" href="#latent-time-trajectories-for-df-selection"></a>
</h2>
<p>Experimentation on metabolic data during the development of <code>santaR</code> confirmed these observations; automated approaches can be suitable when more than 20 data-points are considered, but fail for shorter time-series. This lack of reliability results in over-fitting when each variable is fitted separately, leading to erratic analysis results. It does not seem that automated approaches are currently able to reliably solve the fitting of short time-series, as enough information describing the underlying function’s complexity might not be present in such a limited number of measurements. Ramsay <em>et al</em>.<a href="#fn30" class="footnote-ref" id="fnref30"><sup>30</sup></a> acknowledge this limitation and indicate that <em>“if the data are not telling us all that much about <span class="math inline">\(\lambda\)</span>, then it is surely reasonable to use your judgment in working with values which seem to provide more useful results than the minimizing value does”</em>, highlighting the work of Chaudhuri and Marron <a href="#fn31" class="footnote-ref" id="fnref31"><sup>31</sup></a> <a href="#fn32" class="footnote-ref" id="fnref32"><sup>32</sup></a>.</p>
<p>Chaudhuri and Marron draw a parallel between the bandwidth controlling the smoothness in KDE, and the “resolution” of the problem studied. As the model complexity increases (decreased smoothness, increased <em>df</em>) so does the resolution (in the time domain) at which we try to characterise the underlying function. A higher resolution means observing grainier features, but increases the risk of generating spurious sampling artefacts. A visual exploration of the possible model fits should help detect spurious fit based on a user’s intuition of the resolution provided by the current sampling and knowledge of the system being studied. Automated methods where deemed essential due to possible user inexperience, the need for standardisation and the impracticality of manual annotation of large datasets<a href="#fn33" class="footnote-ref" id="fnref33"><sup>33</sup></a>. However the current failure of automation compromises efforts in standardisation and the use by totally inexperienced user. In turn, if the data is to be presented for visual assessment by a user (knowledgeable on the process under study, and which can appreciate the time resolution provided by the experimental design), a smaller representation of the dataset must be generated.</p>
<p>To reduce the scale of manual annotation, the present strategy proposes to extract and fit the latent time-trajectories across all variables. Using PCA, the data matrix can be decomposed and the most representative temporal profiles (given by the eigenvectors that explain the majority of variance in the data) identified. By inspecting the fitting of these eigen-projections over a range of <em>df</em>, the most suitable and informative smoothing parameter to apply across all variables can then be selected. The underlying assumption is that, <strong>if a satisfactory fit can be obtained on the latent time-trajectories, the selected smoothness should suit all variables</strong>.</p>
<p>Originally devised by Alter <em>et al</em>.<a href="#fn34" class="footnote-ref" id="fnref34"><sup>34</sup></a> and subsequently employed by Storey et al. EDGE<a href="#fn35" class="footnote-ref" id="fnref35"><sup>35</sup></a>, the so called “eigen-genes” (eigen-genes are eigen-vectors in the gene space) are extracted after singular value decomposition (SVD) of the data. The top eigen-genes represent the directions in the gene space that explain the most variance. These applications subsequently employed an automated fit of the top eigen-genes<a href="#fn36" class="footnote-ref" id="fnref36"><sup>36</sup></a> <a href="#fn37" class="footnote-ref" id="fnref37"><sup>37</sup></a> <a href="#fn38" class="footnote-ref" id="fnref38"><sup>38</sup></a>, while we propose to visually select the optimal fit.</p>
<p>While conceptually described, none of these approaches clearly explain how eigen-genes are computed. Alter <em>et al</em>.<a href="#fn39" class="footnote-ref" id="fnref39"><sup>39</sup></a> realise the SVD on a matrix of values for genes (variable) as row and arrays as columns. As arrays correspond to subsequent measurements, they are ordered by time. Each row therefore represents a variable’s time-trajectory. This approach does not address the case where replicates (multiple subjects) have been investigated; when two different experiments were conducted, the second experimental design was added as additional columns. As trajectories are considered as rows, the second experiment is considered as additional time-points of the original experiment condition. The top eigengenes then extract the “interaction” pattern between the two experimental designs instead of the top variance in the first or the second experiment, greatly penalising the interpretation of the results (see Supplementary Information of Alter <em>et al</em>.<a href="#fn40" class="footnote-ref" id="fnref40"><sup>40</sup></a>). EDGE employs a similar methodology, without providing more information on how the case of multiple measurements and multiple individuals is addressed.</p>
<p>Taking inspiration from these approaches, <code>santaR</code> proposes to generate eigen-splines, representing the most representative time-trajectories (i.e. explaining the most variance) in the variable space (Fig. 1).</p>
<p><img src="../reference/figures/santaR-eigenspline.png" width="700px"></p>
<blockquote>
<p>Fig, 1: Schematic representation of the procedure for the extraction of latent time-trajectories (or eigen-trajectories) explaining the most variance in the variable space. The underlying hypothesis is that a satisfactory parametrisation of the fit of these latent time-trajectories should suit the fitting of most trajectories in the dataset. <em>PCA</em>: Principal component analysis, <em>PC</em>: Principal component, <em>Ind</em>: Individual.</p>
</blockquote>
<p>First, as the interest resides in the functional shape across subjects and variables, all observations for a given variable (irrespective of time or subject) are auto-scaled. Auto-scaling corresponds to mean centering and Unit Variance (UV) scaling of the data. Mean centering ensures the mean value of a variable is 0, adjusting for “baseline” differences across variables. UV-scaling adjusts for differences in variance across variables which could bias subsequent PCA. After Auto-scaling, each variable present a standard deviation of one and a mean of zero.</p>
<p>For each variable, a matrix of measured values for each individual (rows) and time-point (columns) is generated. The time employed comprises all the unique time-points across all variables; if no measurement is available for a given time and individual, the cell is left empty. Depending on the regularity of sampling and study design, these matrices can potentially present some sparsity.</p>
<p>All the variable’s data matrix are concatenated vertically. The resulting data matrix contains all the measured values for each variable in each individual (rows) at every observed time-point (columns). Each row now represents the measured values of a variable for a given subject, corresponding to a time-trajectory. As the scale and baseline difference across variables have been removed, the shape information of these trajectories can be compared. Principal component analysis of the transposed data matrix is then executed. The maximal number of principal components is the minimum between the number of time-points (columns) and the number of individual variable measurements (rows), minus one. In practice the number of unique time-point is most often the smallest dimension, resulting in a low number of principal components. Each principal component reflects a mixture of individual variable measurements (loadings) while the scores present a projection of each time-point onto each principal component. For each principal component (representing the direction of variance in the variable space), the score of each time-point can be extracted, resulting in an eigen-trajectory that can be fitted.</p>
<p>Obvious limitations do exist. First, as all variables are considered, the obtained trajectories are dependent on the set of variable provided. Secondly, as the data matrix could potentially be sparse, the PCA algorithm might have to handle missing values, while no measure of success (or warning when assumptions are breached) is available. This could potentially undermine the generation of eigen-trajectories, however we must bear in mind that relying on potential latent description of the underlying data might ultimately be a flawed process. It could be questioned if historically the use of such approaches has not been undertaken to prevent anf automated algorithm from over-fitting a set of time-trajectories.</p>
<p>With a limited set of potential eigen-trajectories to fit (usually the number of unique time-points minus one), <code>santaR</code> can provide multiple measure to help select the most adequate degree of smoothness. The value of <em>df</em> minimising the <strong>CV</strong> score, <strong>GCV</strong> score, <strong>AIC</strong>, <strong>BIC</strong> or <strong>AICc</strong> can be found for each eigen-trajectory. The resulting <em>df</em> values can be averaged across all eigen-trajectories or weighted by the variance explained of each principal component. In practice, the result of automated fitting can vary greatly from an eigen-trajectory to the next (corresponding to under- and over-fitting), with no consensus appearing across all components. Additionally no single metric outperforms the others in quality of fit and reliability.</p>
<p>Trajectories can also be visually inspected for varying <em>df</em>, while taking into account a last factor; the number of time-points <em>N</em> must be superior or equal to <em>df</em> in order to fit a smoothing spline. In other terms, for a high number of degree of freedom and missing data, a substantial number of trajectories might have to be rejected. In practice most model selection algorithms tend to over-fit the data and require discarding most metabolic trajectories.</p>
</div>
<div class="section level2">
<h2 id="smoothness-and-experimental-design">Smoothness and experimental design<a class="anchor" aria-label="anchor" href="#smoothness-and-experimental-design"></a>
</h2>
<p>The smoothness of a time-trajectory should be determined from the noisy realisation of the underlying function of time. Yet, such data driven approaches often fail to balance the goodness-of-fit of the data versus the models complexity. With too few informative measurements available, the limitations of automated smoothness selection undermines the algorithm, resulting in unreliable fit of the data and over-fitted results. A first solution to this challenge is to assume that a single trend is underlying all variables’ observations, and therefore a single degree of smoothness should be applied across all time-trajectories. Based on this assumption and the expectation that an informed user will intuitively achieve a more consistent fit of the data than an automated algorithm on short noisy time-series; <code>santaR</code> considers <em>df</em> as the single meta-parameter a user must choose prior to analysis, balancing the fitting of raw observations versus the smoothing of the biological variability and measurements errors. In turn this explicit selection of the model complexity increases the reproducibility of results and reduces the execution time.</p>
<p>To help visualise the time-trajectories present in a dataset and make an informed judgment, we presented an approach based on PCA to extract and fit latent time-trajectories across all variables. While multiple goodness-of-fit measures are proposed to help guide <em>df</em> selection, some useful pointers can be highlighted. First, <em>df</em> should not be too elevated, as to not have to reject too many trajectories (<em>df &gt; N</em>), but also as to not generate interpolation splines (<em>df = N</em>). Two data points can always define a line, three points a parabola and four a cubic function; however, for a meaningful fit, the number of measurements should exceed the minimum number of points required to define a shape. Secondly, based on the work of Chaudhuri and Marron<a href="#fn41" class="footnote-ref" id="fnref41"><sup>41</sup></a> <a href="#fn42" class="footnote-ref" id="fnref42"><sup>42</sup></a> <strong>smoothing can be thought as a problematic of resolution; the question is not how many time-points or even the range of time covered, but what could be the functional trend properly characterised with the current sampling frequency</strong>.</p>
<p>If we take the example of a patient followed over 5 days, it can reasonably be expected that his metabolic profile will reflect a degradation of his condition, and maybe an amelioration. One measurement per day, equalling 6 time-points could reflect such a function of time, containing one (or maximum two) inflection. If the question was to characterise the circadian evolution of a metabolite on top of the “long term” evolution of the patient (a “high resolution” question), the present sampling is not sufficient. A slightly more adequate study design to resolve this question would require a measurement around every 4 to 6 hours, resulting in 20 to 30 observations( conversely, for a “low resolution” event like disease progression, such a high sampling frequency is not necessary, albeit it could help fitting) (Fig. 2).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## ----------------------------------------------------</span>
<span class="co">## 2 levels of resolution</span>
<span class="co">## ----------------------------------------------------</span>

<span class="co"># make curve (cos) with added circadian variation (sin)</span>
<span class="va">x1</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.5</span>, <span class="fl">5.5</span>, <span class="fl">0.01</span><span class="op">)</span>
<span class="va">x11</span> <span class="op">=</span> <span class="op">(</span><span class="va">x1</span><span class="op">*</span><span class="op">(</span><span class="fl">2</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span><span class="op">*</span><span class="va">pi</span><span class="op">)</span><span class="op">+</span><span class="va">pi</span>
<span class="va">y1</span>  <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">x11</span><span class="op">+</span><span class="op">(</span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">x1</span><span class="op">*</span><span class="fl">2</span><span class="op">*</span><span class="va">pi</span><span class="op">)</span> <span class="op">+</span><span class="fl">3.5</span>

<span class="co"># one measurement per day</span>
<span class="va">x2</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">x22</span> <span class="op">=</span> <span class="op">(</span><span class="va">x2</span><span class="op">*</span><span class="op">(</span><span class="fl">2</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span><span class="op">*</span><span class="va">pi</span><span class="op">)</span><span class="op">+</span><span class="va">pi</span>
<span class="va">y2</span>  <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">x22</span><span class="op">+</span><span class="op">(</span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">x2</span><span class="op">*</span><span class="fl">2</span><span class="op">*</span><span class="va">pi</span><span class="op">)</span> <span class="op">+</span><span class="fl">3.5</span>

<span class="co"># one measurement every 4 hours</span>
<span class="va">x3</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span>
<span class="va">x33</span> <span class="op">=</span> <span class="op">(</span><span class="va">x3</span><span class="op">*</span><span class="op">(</span><span class="fl">2</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span><span class="op">*</span><span class="va">pi</span><span class="op">)</span><span class="op">+</span><span class="va">pi</span>
<span class="va">y3</span>  <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">x33</span><span class="op">+</span><span class="op">(</span><span class="op">(</span><span class="va">pi</span><span class="op">)</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">x3</span><span class="op">*</span><span class="fl">2</span><span class="op">*</span><span class="va">pi</span><span class="op">)</span> <span class="op">+</span><span class="fl">3.5</span>


<span class="co">## generate a spline at each resolution------------------------------</span>
<span class="va">time</span>      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">0.01</span><span class="op">)</span>
<span class="va">tmp_fit0</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/smooth.spline.html" class="external-link">smooth.spline</a></span><span class="op">(</span> <span class="va">x1</span>, <span class="va">y1</span>, df<span class="op">=</span><span class="fl">601</span><span class="op">)</span> <span class="co"># original</span>
<span class="va">tmp_fit1</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/smooth.spline.html" class="external-link">smooth.spline</a></span><span class="op">(</span> <span class="va">x2</span>, <span class="va">y2</span>, df<span class="op">=</span><span class="fl">6</span><span class="op">)</span>   <span class="co"># daily sampling</span>
<span class="va">tmp_fit2</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/smooth.spline.html" class="external-link">smooth.spline</a></span><span class="op">(</span> <span class="va">x3</span>, <span class="va">y3</span>, df<span class="op">=</span><span class="fl">31</span><span class="op">)</span>  <span class="co"># 4h sampling</span>


<span class="co"># plot each time trajectory------------------------------------------</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span>

<span class="va">tmpDaily</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> x<span class="op">=</span><span class="va">x2</span>,  y<span class="op">=</span><span class="va">y2</span><span class="op">)</span>
<span class="va">tmp4h</span>     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> x<span class="op">=</span><span class="va">x3</span>,  y<span class="op">=</span><span class="va">y3</span><span class="op">)</span>

<span class="va">pred0</span>    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span> object<span class="op">=</span><span class="va">tmp_fit0</span>, x<span class="op">=</span><span class="va">time</span> <span class="op">)</span>
<span class="va">pred1</span>    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span> object<span class="op">=</span><span class="va">tmp_fit1</span>, x<span class="op">=</span><span class="va">time</span> <span class="op">)</span>
<span class="va">pred2</span>    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span> object<span class="op">=</span><span class="va">tmp_fit2</span>, x<span class="op">=</span><span class="va">time</span> <span class="op">)</span>

<span class="va">tmpPred0</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> x<span class="op">=</span><span class="va">pred0</span><span class="op">$</span><span class="va">x</span>, y<span class="op">=</span><span class="va">pred0</span><span class="op">$</span><span class="va">y</span><span class="op">)</span>
<span class="va">tmpPred1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> x<span class="op">=</span><span class="va">pred1</span><span class="op">$</span><span class="va">x</span>, y<span class="op">=</span><span class="va">pred1</span><span class="op">$</span><span class="va">y</span><span class="op">)</span>
<span class="va">tmpPred2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> x<span class="op">=</span><span class="va">pred2</span><span class="op">$</span><span class="va">x</span>, y<span class="op">=</span><span class="va">pred2</span><span class="op">$</span><span class="va">y</span><span class="op">)</span>

<span class="va">p0</span>      <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, environment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">environment</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">7.1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">p0</span>      <span class="op">&lt;-</span> <span class="va">p0</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span> data<span class="op">=</span><span class="va">tmpPred0</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span>, linetype<span class="op">=</span><span class="fl">1</span>, color<span class="op">=</span><span class="st">'black'</span> <span class="op">)</span>
<span class="va">p0</span>      <span class="op">&lt;-</span> <span class="va">p0</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">tmp4h</span>,    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">2.5</span>,   color<span class="op">=</span><span class="st">'grey60'</span> <span class="op">)</span>
<span class="va">p0</span>      <span class="op">&lt;-</span> <span class="va">p0</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">tmpDaily</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">2.5</span>,   color<span class="op">=</span><span class="st">'grey25'</span> <span class="op">)</span>
<span class="va">p0</span>      <span class="op">&lt;-</span> <span class="va">p0</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Variable response'</span><span class="op">)</span>

<span class="va">p2</span>      <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, environment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">environment</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">7.1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">p2</span>      <span class="op">&lt;-</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">tmp4h</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">2.5</span>, color<span class="op">=</span><span class="st">'springgreen3'</span> <span class="op">)</span>
<span class="va">p2</span>      <span class="op">&lt;-</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span> data<span class="op">=</span><span class="va">tmpPred2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span>, linetype<span class="op">=</span><span class="fl">1</span>, color<span class="op">=</span><span class="st">'springgreen3'</span> <span class="op">)</span>
<span class="va">p2</span>      <span class="op">&lt;-</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">tmpDaily</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">2.5</span>, color<span class="op">=</span><span class="st">'mediumblue'</span> <span class="op">)</span>
<span class="va">p2</span>      <span class="op">&lt;-</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span> data<span class="op">=</span><span class="va">tmpPred1</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span>, linetype<span class="op">=</span><span class="fl">1</span>, color<span class="op">=</span><span class="st">'mediumblue'</span> <span class="op">)</span>
<span class="va">p2</span>      <span class="op">&lt;-</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Time'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Variable response'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">p0</span>,<span class="va">p2</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="selecting-optimal-df_files/figure-html/unnamed-chunk-2-1.png" width="672"></p>
<blockquote>
<p>Fig. 2: Sampling of a biological perturbation at multiple resolutions. The true smooth evolution of a variable is observed over 5 days (top, <strong>black line</strong>: real evolution, <strong>black points</strong>: low resolution measurements, <strong>grey points</strong>: high resolution measurements), representing the combination of the perturbation of a metabolite on top of its usual circadian rhythm.</p>
<ul>
<li>A daily sampling, (6 measurements) enable a general approximation of the metabolite evolution but not of the circadian variation (Bottom, blue line).</li>
<li>A higher sampling frequency over the same timeframe (every 4 hours) enable the additional description of the circadian variations of the metabolite, corresponding to a more complex (higher resolution) time evolution (bottom, green line ).</li>
</ul>
</blockquote>
<p>In other words, <em>df</em> is ultimately a function of study design, relating to the number of time-points, the sampling rate, and most importantly the complexity of the underlying function of time (Fig. 3). While a data-driven approach may seem appealing, an algorithm currently cannot infer this information and only the user can provide it.</p>
<p><img src="../reference/figures/santaR-complexity_and_sampling.png" width="500px"></p>
<blockquote>
<p>Fig. 3: Visualisation of the adequate sampling rate (study design) in light of the complexity of the expected function of time.</p>
<p>Four simple and two complex functions of time are sampled (with error) at a maximum of five time-points (red dots).</p>
<p>While sufficient to satisfactorily approximate a simple function, this number of measurements must be increased (grey dots) for more complex time-trajectories such as the longitudinal sampling of a patient along a surgical journey (with possible recovery and further complications). The use of only five time-points for such a complex trajectory would require perfectly timed measurements, without errors and a subsequent over-fitting of the observations.</p>
</blockquote>
<p>Without going as far as Hodrick and Prescott<a href="#fn43" class="footnote-ref" id="fnref43"><sup>43</sup></a> <a href="#fn44" class="footnote-ref" id="fnref44"><sup>44</sup></a> which advocated that the smoothing parameter <span class="math inline">\(\lambda\)</span> should always take a particular value (1600) regardless of the data<a href="#fn45" class="footnote-ref" id="fnref45"><sup>45</sup></a>, it can be noted a similarity in the <em>df</em> employed by multiple methods when applied to short biological time-series. The applications of EDGE found the optimal dimensionality of the spline basis to be similar across experiments<a href="#fn46" class="footnote-ref" id="fnref46"><sup>46</sup></a>. Likewise, for experiments containing 5-10 time points and investigating functions of time presenting one or two inflections (as is the limit in resolution of such a sized experiment), 4 to 6 degrees of freedom seem to be a good starting point to suitably describe the time-trajectory without over-fitting, interpolating or excessively rejecting trajectories . Finally, an increase of sampling frequency does not automatically imply a more complex model should be fitted; if the expected functional form over the defined time-frame is “simple”, the model complexity should reflect it.</p>
</div>
<div class="section level2">
<h2 id="see-also">See Also<a class="anchor" aria-label="anchor" href="#see-also"></a>
</h2>
<ul>
<li><a href="getting-started.html">Getting Started with santaR</a></li>
<li><a href="prepare-input-data.html">How to prepare input data for santaR</a></li>
<li><a href="theoretical-background.html">santaR theoretical background</a></li>
<li><a href="santaR-GUI.pdf">Graphical user interface use</a></li>
<li><a href="automated-command-line.html">Automated command line analysis</a></li>
<li><a href="plotting-options.html">Plotting options</a></li>
<li><a href="advanced-command-line-functions.html">Advanced command line options</a></li>
</ul>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Déjean, S., Martin, P. G. P., Baccini, A. &amp; Besse, P. <em>Clustering time-series gene expression data using smoothing spline derivatives</em>. Eurasip Journal on Bioinformatics and Systems Biology 2007, 10 (2007)<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Ramsay, J. &amp; Silverman, B. W. <em>Functional Data Analysis</em> Springer, 431 (John Wiley &amp; Sons, Ltd, Chichester, UK, 2005)<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Chaudhuri, P. &amp; Marron, J. S. <em>SiZer for Exploration of Structures in Curves</em>. Journal of the American Statistical Association 94, 807-823 (1999)<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>Chaudhuri, P. &amp; Marron, J. S. <em>Scale Space View of Curve Estimation</em>. The Annals of Statistics 28, 408-428 (2000)<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>Hastie, T., Tibshirani, R. &amp; Friedman, J. <em>The Elements of Statistical Learning</em>. (Springer, 2009)<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p>Craven, P. &amp; Wahba, G. <em>Smoothing noisy data with spline functions - Estimating the correct degree of smoothing by the method of generalized cross-validation</em>. Numerische Mathematik 31, 377-403 (1978)<a href="#fnref6" class="footnote-back">↩︎</a></p></li>
<li id="fn7"><p>Hastie, T., Tibshirani, R. &amp; Friedman, J. <em>The Elements of Statistical Learning</em>. (Springer, 2009)<a href="#fnref7" class="footnote-back">↩︎</a></p></li>
<li id="fn8"><p>Ramsay, J., Hooker, G. &amp; Graves, S. <em>Functional data analysis with R and MATLAB</em> (Springer-Verlag, 2009)<a href="#fnref8" class="footnote-back">↩︎</a></p></li>
<li id="fn9"><p>Rice, J. A. &amp; Wu, C. O. <em>Nonparametric mixed effects models for unequally sampled noisy curves</em>. Biometrics 57, 253-259 (2001)<a href="#fnref9" class="footnote-back">↩︎</a></p></li>
<li id="fn10"><p>James, G. M. &amp; Hastie, T. J. <em>Functional linear discriminant analysis for irregularly sampled curves</em>. Journal of the Royal Statistical Society, Series B 63, 533-550 (2001)<a href="#fnref10" class="footnote-back">↩︎</a></p></li>
<li id="fn11"><p>Akaike, H. <em>Information Theory And An Extension Of The Maximum Likelihood Principle By Hirotogu Akaike</em>. Second International Symposium on Information Theory, 267-281 (1973)<a href="#fnref11" class="footnote-back">↩︎</a></p></li>
<li id="fn12"><p>Akaike, H. <em>A New Look at the Statistical Model Identification</em>. IEEE Transactions on Automatic Control 19, 716-723 (1974)<a href="#fnref12" class="footnote-back">↩︎</a></p></li>
<li id="fn13"><p>Schwarz, G. <em>Estimating the Dimension of a Model</em>. The Annals of Statistics 6, 461-464 (1978)<a href="#fnref13" class="footnote-back">↩︎</a></p></li>
<li id="fn14"><p>Wu, H. &amp; Zang, J.-T. <em>Nonparametric Regression Methods for Longitudinal Data Analysis: mixed-effects modeling approaches</em> (Wiley-VCH Verlag GmbH &amp; Co. KGaA, Hoboken, NJ, USA, 2006)<a href="#fnref14" class="footnote-back">↩︎</a></p></li>
<li id="fn15"><p>Hedeker, D. R. &amp; Gibbons, R. D. <em>Longitudinal Data Analysis</em> (Wiley-Interscience, 2006)<a href="#fnref15" class="footnote-back">↩︎</a></p></li>
<li id="fn16"><p>Azari, R., Li, L. &amp; Tsai, C.-L. <em>Longitudinal data model selection</em>. Computational Statistics &amp; Data Analysis 50, 3053-3066 (2006)<a href="#fnref16" class="footnote-back">↩︎</a></p></li>
<li id="fn17"><p>Hastie, T., Tibshirani, R. &amp; Friedman, J. <em>The Elements of Statistical Learning</em>. (Springer, 2009)<a href="#fnref17" class="footnote-back">↩︎</a></p></li>
<li id="fn18"><p>Hurvich, C. M. &amp; Tsai, C.-L. <em>Regression and time series model selection in small samples</em>. Biometrika 76, 297-307 (1989)<a href="#fnref18" class="footnote-back">↩︎</a></p></li>
<li id="fn19"><p>Bar-Joseph, Z., Gerber, G., Simon, I., Gifford, D. K. &amp; Jaakkola, T. S. <em>Comparing the continuous representation of time-series expression profiles to identify differentially expressed genes</em>. Proceedings of the National Academy of Sciences of the United States of America 100, 10146-10151 (2003)<a href="#fnref19" class="footnote-back">↩︎</a></p></li>
<li id="fn20"><p>Ernst, J., Nau, G. J. &amp; Bar-Joseph, Z. <em>Clustering short time series gene expression data</em>. Bioinformatics 21, 159-168 (2005)<a href="#fnref20" class="footnote-back">↩︎</a></p></li>
<li id="fn21"><p>Storey, J. D., Xiao, W., Leek, J. T., Tompkins, R. G. &amp; Davis, R. W. <em>Significance analysis of time course microarray experiments</em>. Proceedings of the National Academy of Sciences of the United States of America 102, 12837-42 (2005)<a href="#fnref21" class="footnote-back">↩︎</a></p></li>
<li id="fn22"><p>Berk, M., Ebbels, T. M. D. &amp; Montana, G. <em>A statistical framework for biomarker discovery in metabolomic time course data</em>. Bioinformatics (Oxford, England) 27, 1979-85 (2011)<a href="#fnref22" class="footnote-back">↩︎</a></p></li>
<li id="fn23"><p>Ramsay, J., Hooker, G. &amp; Graves, S. <em>Functional data analysis with R and MATLAB</em> (Springer-Verlag, 2009)<a href="#fnref23" class="footnote-back">↩︎</a></p></li>
<li id="fn24"><p>Déjean, S., Martin, P. G. P., Baccini, A. &amp; Besse, P. <em>Clustering time-series gene expression data using smoothing spline derivatives</em>. Eurasip Journal on Bioinformatics and Systems Biology 2007, 10 (2007)<a href="#fnref24" class="footnote-back">↩︎</a></p></li>
<li id="fn25"><p>Lee, T. C. M. <em>Smoothing parameter selection for smoothing splines: A simulation study</em>. Computational Statistics and Data Analysis 42, 139-148 (2003)<a href="#fnref25" class="footnote-back">↩︎</a></p></li>
<li id="fn26"><p>Ramsay, J., Hooker, G. &amp; Graves, S. _Functional data analysis with R and MATLAB (Springer-Verlag, 2009)<a href="#fnref26" class="footnote-back">↩︎</a></p></li>
<li id="fn27"><p>Ramsay, J. O. &amp; Silverman, B. <em>Applied functional data analysis: methods and case studies</em> (eds Ramsay, J. O. &amp; Silverman, B. W.) (Springer New York, New York, NY, 2002)<a href="#fnref27" class="footnote-back">↩︎</a></p></li>
<li id="fn28"><p>Storey, J. D., Xiao, W., Leek, J. T., Tompkins, R. G. &amp; Davis, R. W. <em>Significance analysis of time course microarray experiments</em>. Proceedings of the National Academy of Sciences of the United States of America 102, 12837-42 (2005)<a href="#fnref28" class="footnote-back">↩︎</a></p></li>
<li id="fn29"><p>Déjean, S., Martin, P. G. P., Baccini, A. &amp; Besse, P. <em>Clustering time-series gene expression data using smoothing spline derivatives</em>. Eurasip Journal on Bioinformatics and Systems Biology 2007, 10 (2007)<a href="#fnref29" class="footnote-back">↩︎</a></p></li>
<li id="fn30"><p>Ramsay, J., Hooker, G. &amp; Graves, S. <em>Functional data analysis with R and MATLAB</em> (Springer-Verlag, 2009)<a href="#fnref30" class="footnote-back">↩︎</a></p></li>
<li id="fn31"><p>Chaudhuri, P. &amp; Marron, J. S. <em>SiZer for Exploration of Structures in Curves</em>. Journal of the American Statistical Association 94, 807-823 (1999)<a href="#fnref31" class="footnote-back">↩︎</a></p></li>
<li id="fn32"><p>Chaudhuri, P. &amp; Marron, J. S. <em>Scale Space View of Curve Estimation</em>. The Annals of Statistics 28, 408-428 (2000)<a href="#fnref32" class="footnote-back">↩︎</a></p></li>
<li id="fn33"><p>Silverman, B. <em>Some Aspects of the Spline Smoothing Approach to Non-Parametric Regresion Curve Fitting</em>. Journal of the Royal Statistical Society, Series B 47, 1-52 (1985)<a href="#fnref33" class="footnote-back">↩︎</a></p></li>
<li id="fn34"><p>Alter, O., Brown, P. O. &amp; Botstein, D. <em>Singular value decomposition for genome-wide expression data processing and modeling</em>. Proceedings of the National Academy of Sciences of the United States of America 97, 10101-10106 (2000)<a href="#fnref34" class="footnote-back">↩︎</a></p></li>
<li id="fn35"><p>Storey, J. D., Xiao, W., Leek, J. T., Tompkins, R. G. &amp; Davis, R. W. <em>Significance analysis of time course microarray experiments</em>. Proceedings of the National Academy of Sciences of the United States of America 102, 12837-42 (2005)<a href="#fnref35" class="footnote-back">↩︎</a></p></li>
<li id="fn36"><p>Storey, J. D., Xiao, W., Leek, J. T., Tompkins, R. G. &amp; Davis, R. W. <em>Significance analysis of time course microarray experiments</em>. Proceedings of the National Academy of Sciences of the United States of America 102, 12837-42 (2005)<a href="#fnref36" class="footnote-back">↩︎</a></p></li>
<li id="fn37"><p>Alter, O., Brown, P. O. &amp; Botstein, D. <em>Singular value decomposition for genome-wide expression data processing and modeling</em>. Proceedings of the National Academy of Sciences of the United States of America 97, 10101-10106 (2000)<a href="#fnref37" class="footnote-back">↩︎</a></p></li>
<li id="fn38"><p>Hong, F. &amp; Li, H. <em>Functional hierarchical models for identifying genes with different time-course expression profiles</em>. Biometrics 62, 534-544 (2006)<a href="#fnref38" class="footnote-back">↩︎</a></p></li>
<li id="fn39"><p>Alter, O., Brown, P. O. &amp; Botstein, D. <em>Singular value decomposition for genome-wide expression data processing and modeling</em>. Proceedings of the National Academy of Sciences of the United States of America 97, 10101-10106 (2000)<a href="#fnref39" class="footnote-back">↩︎</a></p></li>
<li id="fn40"><p>Alter, O., Brown, P. O. &amp; Botstein, D. <em>Singular value decomposition for genome-wide expression data processing and modeling</em>. Proceedings of the National Academy of Sciences of the United States of America 97, 10101-10106 (2000)<a href="#fnref40" class="footnote-back">↩︎</a></p></li>
<li id="fn41"><p>Chaudhuri, P. &amp; Marron, J. S. <em>SiZer for Exploration of Structures in Curves</em>. Journal of the American Statistical Association 94, 807-823 (1999)<a href="#fnref41" class="footnote-back">↩︎</a></p></li>
<li id="fn42"><p>Chaudhuri, P. &amp; Marron, J. S. <em>Scale Space View of Curve Estimation</em>. The Annals of Statistics 28, 408-428 (2000)<a href="#fnref42" class="footnote-back">↩︎</a></p></li>
<li id="fn43"><p>Hodrick, R. J. &amp; Prescott, E. C. <em>Postwar U.S. Business Cycles: An Empirical Investigation</em>. Carnegie Mellon University discussion paper (1980)<a href="#fnref43" class="footnote-back">↩︎</a></p></li>
<li id="fn44"><p>Hodrick, R. J. &amp; Prescott, E. C. <em>Postwar US Business Cycles : An Empirical Investigation.</em> Journal of Money, Credit and Banking 29, 1-16 (1997)<a href="#fnref44" class="footnote-back">↩︎</a></p></li>
<li id="fn45"><p>Paige, R. L. &amp; Trindade, A. A. <em>The Hodrick-Prescott Filter: A special case of penalized spline smoothing</em>. Electronic Journal of Statistics 4, 856-874 (2010).<a href="#fnref45" class="footnote-back">↩︎</a></p></li>
<li id="fn46"><p>Storey, J. D., Xiao, W., Leek, J. T., Tompkins, R. G. &amp; Davis, R. W. <em>Supporting Appendix Significance analysis of time course microarray experiments</em>. 1-27<a href="#fnref46" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Arnaud Wolfer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
